source/mainnet/smart-contracts/best-practices/costs.rst

```diff
@@ -0,0 +1,75 @@
+.. _sc-costs-best-practices:
+
+=============================
+Cost reduction best practices
+=============================
+
+There are several ways that you can reduce transaction costs when developing your smart contracts.
+
+Use a custom allocator
+======================
+
+Smart contracts sometimes need to dynamically allocate memory.
+A "memory allocator" program performs this job.
+Rust programs compiled with default compiler settings use the
+`standard allocator <https://doc.rust-lang.org/std/alloc/struct.System.html>`_
+implemented in the Rust standard library.
+However, you can also specify a custom allocator.
+The recommended option is ``bump_alloc`` which ships with ``concordium-std >= 10.0.0``.
+To enable it, you must enable the feature ``bump_alloc``:
+
+.. code-block:: rust
+
+    [dependencies]
+    concordium-std = {version = "10.0", features = ["bump_alloc"]}
+
+Alternatively, if you want to test with and without ``bump_alloc`` enabled, add a ``bump_alloc`` feature to the smart contract crate as follows:
+
+.. code-block:: rust
+
+    [features]
+    default = ["std", "bump_alloc"]
+    std = ["concordium-std/std"]
+    bump_alloc = ["concordium-std/bump_alloc"]
+
+While enabling a custom allocator is optional by default, it is required if you are using the ``no-std`` feature in your smart contract, as that disables the default allocator.
+
+The main reason for using ``bump_alloc`` instead of the default allocator,
+even in ``std`` builds, is that ``bump_alloc`` has a smaller code footprint,
+i.e, the resulting smart contracts are going to be smaller by about 6-10kB,
+which means they are cheaper to deploy and run. ``bump_alloc`` is designed to
+be simple and fast, but it does not use the memory very efficiently. For
+short-lived programs, such as smart contracts, this is usually the right
+tradeoff. Especially for contracts such as those dealing with tokens.
+For very complex contracts it may be beneficial to run benchmarks to see
+whether ``bump_alloc`` is the best option. See the Rust `allocator <https://doc.rust-lang.org/std/alloc/index.html#the-global_allocator-attribute>`_
+documentation for more context and details on using custom allocators.
+
+There are other allocators available, for example ``dlmalloc``, that can be used instead.
+
+Use State-* types for maps, sets, and boxes
+===========================================
+
+If your contract uses maps, sets, or boxes, then *strongly* consider using the alternatives prefixed with ``State`` from ``concordium-std``. For example |StateMap|_, |StateBTreeMap|_, |StateSet|_, |StateBTreeSet|_, or |StateBox|_.
```

Comment:
Maybe a minor detail, but they are probably only more cost-efficient if you need to persist this collection across calls and don't need to read the entire collection on every call, otherwise I would expect rust-std types to be more cost-efficient, at least in some scenarios :shrug: 
But I believe this is true for a majority of the use cases, so I am okay with adding this

Reply:
Good point, I've added "and stores"